lib/selections.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * @typedef {Object} MinimalSelectionItem
+ * @property {number} from
+ * @property {number} to
+ */
+
+/** A selection of pieces to download.
+ * @typedef {MinimalSelectionItem & {
+ *  offset: number,
+ *  priority: number,
+ *  notify?: function
+ *  isStreamSelection?: boolean
+ * }} SelectionItem
+ */
+
+/**
+ * @typedef {MinimalSelectionItem & {notify: function}} NotificationItem
+ */
+
+export class Selections {
+  /** @type {Array<SelectionItem>} */
+  _items = []
+
+  /**
+   * @param {MinimalSelectionItem & {isStreamSelection?: boolean}} item Interval to be removed from the selection
+   */
+  remove (item) {
+    for (let i = 0; i < this._items.length; i++) {
+      const existing = this._items[i]
+      if (existing.isStreamSelection) {
+        if (item.isStreamSelection) {
+          // If both are stream selections and they match, then we remove the first matching item, then we break the loop
+          if (existing.from === item.from && existing.to === item.to) {
+            this._items.splice(i, 1)
+            // for stream selections, we only remove one item at a time
+            // ergo we break the loop after removing the first matching item
+            break
+          }
+        } else {
+          // we only remove stream selections when the `isStreamSelection` flag is true and they match
+          continue
+        }
+      } else {
+        if (isLowerIntersecting(item, existing)) {
+          existing.to = Math.max(item.from - 1, 0)
+        } else if (isUpperIntersecting(item, existing)) {
+          existing.from = item.to + 1
+        } else if (isInsideExisting(item, existing)) {
+          const replacingItems = []
+          const existingStart = { ...existing, to: Math.max(item.from - 1, 0) }
+          if (existingStart.to - existingStart.from > 0) replacingItems.push(existingStart)
+          const existingEnd = { ...existing, from: item.to + 1 }
+          if (existingEnd.to - existingEnd.from > 0) replacingItems.push(existingEnd)
+          this._items.splice(i, 1, ...replacingItems)
+          i = i - 1 + replacingItems.length // decrement i to offset splice
+        } else if (isCoveringExisting(item, existing)) {
+          this._items.splice(i, 1)
+          i--
+        }
+      }
+    }
+  }
+
+  /**
+   * @param {SelectionItem & NotificationItem} newItem
+   */
+  insert (newItem) {
+    if (newItem.from > newItem.to) {
+      throw new Error('Invalid interval')
+    }
+    if (!newItem.isStreamSelection) {
+      const { notify: oldNotify } = newItem
+      // Merge notifications of intersecting items into the new item's notify function
```

Comment:
shouldn't this entire thing be skipped if notify doesn't exist on the item or am I missunderstanding?

Reply:
i don't think so. even though the inserted item doesn't have a notify function, if the merged items do, then we should merge those into the inserted one.